### PR TITLE
✨ REDmod Layouts For Native Installs

### LINKED ISSUE #244: Cover Remaining REDmod Regular Install Cases
A couple things still missing from plain REDmods

- [ ] Explicit support for all subtypes
- [ ] Tests for all
- [ ] info.json validation tests (e.g. path = name)

### PR DESCRIPTION
**Pull request by effs**
- [x] Support each REDmod layout (Canon, Named, Toplevel)
  - [ ] WONTFIX (for now) TBD is the Toplevel something we really want to support
- [x] Explicitly handle all REDmod subtypes (archives, scripts, tweaks, sounds)
- [x] Validate info.json
  - [x] Currently only validated in Toplevel because we need to get the module name, need to do others  
  - [x] Generally valid, has name & version
  - [x] Name === dir
  - [x] Check sounds defined if sounds? (opt)

- Added io-ts as dependency, provides support for constructing validating decoders (here for info.json but can and should be used for e.g. the AMM JSON as well)

closes #244

### CODE REVIEW

**Review by E1337Kat — COMMENTED** [review]
Looks good so far

File: test/unit/mods.example.redmod.ts [hunk 1]
```diff
@@ -8,53 +11,177 @@ import {
...
+const mySoundModWithoutCustomSoundsFieldJson = JSON.stringify({
+  name: `myRedMod`,
+  version: `1.0.0`,
+  description: `For some reason I left out the required customSounds field`,
+});
+
+const NOTmyRedModInfoJson = JSON.stringify({
+  name: `NOTmyRedMod`,
+  version: `1.0.0`,
+});
```
[the remark this answers is not in the document]
Comment (effs) [reply test/unit/mods.example.redmod.ts:41]: Ye, added now!

File: src/installers.layouts.ts [hunk 2]
```diff
@@ -678,28 +679,85 @@ export const RED4EXT_KNOWN_NONOVERRIDABLE_DLL_DIRS = [path.join(`bin\\x64\\`)];
...
+  Canon = `
+          One or more mods in the form of
+
+          - .\\mods\\[modname]\\info.json { name: modname, ... }
+          - .\\mods\\[modname]\\archives\\*.archive
+          - .\\mods\\[modname]\\customSounds\\*.archive
+          - .\\mods\\[modname]\\scripts\\*.script
+          - .\\mods\\[modname]\\scripts\\**\\*.script
+          - .\\mods\\[modname]\\tweaks\\*.tweak
+          `,
```
[the remark this answers is not in the document]
Comment (effs) [reply src/installers.layouts.ts:699]: Not sure if this makes sense, but if it works, we could

File: test/unit/mods.example.redmod.ts [hunk 3]
```diff
@@ -394,6 +394,61 @@ const REDmodDirectFailures = new Map<string, ExampleFailingMod>([
...
+          myRedMod: {
+            [REDMOD_INFO_FILENAME]: NOTmyRedModInfoJson,
+          },
+        },
+      ),
+      inFiles: [
+        path.join(`info.json`),
+        path.join(`archives/`),
+        path.join(`archives/cool_stuff.archive`),
+      ],
```
[the remark this answers is not in the document]
Comment (effs) [reply test/unit/mods.example.redmod.ts]: Yup, fixed!

File: src/installer.redmod.ts [hunk 4]
```diff
@@ -67,13 +75,50 @@ import { Features } from "./features";
...
+  relativeREDmodDirs: readonly string[],
+  semanticValidations: ((file: File, redmodInfo: REDmodInfo) => Either<Error, REDmodInfo>)[],
+): Promise<Either<Error, readonly ValidInfoFile[]>> =>
   pipe(
-    fileFromDiskTE(pathOnDisk, relativePath),
-    chainEitherKW((file) => J.parse(file.content)),
-    chainEitherKW((json) => REDmodInfoType.decode(json)),
-    mapLeft((err) => new Error(`Error parsing ${relativePath}: ${err}`)),
+    relativeREDmodDirs,
+    traverse(ApplicativePar)((relative) =>
```
[the remark this answers is not in the document]
Comment (effs) [reply src/installer.redmod.ts]: You'll love the new ones then XD

File: test/unit/mods.example.redmod.ts [hunk 5]
```diff
@@ -449,10 +517,137 @@ const REDmodDirectFailures = new Map<string, ExampleFailingMod>([
...
+      inFiles: [
+        path.join(`info.json`),
+        path.join(`archives/`),
+        path.join(`archives/cool_stuff.archive`),
+        path.join(`customSounds/`),
+        path.join(`customSounds/cool_sound.wav`),
+        path.join(`scripts/`),
+        path.join(`scripts/exec/`),
+        path.join(`scripts/exec/cool_scripts.script`),
+        path.join(`scripts/core/ai/`),
```
Comment (E1337Kat) [inline test/unit/mods.example.redmod.ts:612]:
```suggestion
        path.join(`scripts/core/`),
        path.join(`scripts/core/ai/`),
```
path hierarchy

File: src/installer.redmod.ts [hunk 6]
```diff
@@ -167,130 +209,263 @@ export const detectREDmodLayout = (fileTree: FileTree): boolean =>
...
-export const toplevelREDmodLayout = async (
-  api: VortexApi,
-  _modName: string,
-  fileTree: FileTree,
-  installingDir: string,
-): Promise<MaybeInstructions> => {
-  if (!detectToplevelREDmodLayout(fileTree)) {
-    return NoInstructions.NoMatch;
+
+const customSoundLayoutAndValidation = (
```
Comment (E1337Kat) [inline src/installer.redmod.ts:323]: I like this. Always good to validate if we can.

File: src/installer.redmod.ts [hunk 7]
```diff
@@ -313,50 +488,76 @@ export const testForREDmod: V2077TestFunc = (
...
-    return Promise.reject(new Error(errorMessage));
-  }
+  // At this point we have to break out to interop with the rest..
+  const allInstructionsForEverySubmodInside = await allModsForLayoutPipeline();
 
-  return Promise.resolve({ instructions: selectedInstructions.instructions });
+  return isLeft(allInstructionsForEverySubmodInside)
+    ? failAfterWarningUserAndLogging(api, fileTree, modInfo, features, allInstructionsForEverySubmodInside.left)
+    : returnInstructionsAndLogEtc(api, fileTree, modInfo, features, allInstructionsForEverySubmodInside.right);
 };
```
Comment (E1337Kat) [inline src/installer.redmod.ts:563]:
```suggestion
};

```
Maybe it is the ruby dev in me, but I like to have the empty last line for some reason. No need to enforce it though, just commenting here

**Review by E1337Kat — CHANGES_REQUESTED** [review]
more good changes. not sure if the path hierarchy stuff needs to be explicit or not... I feel like it should be so that it is in line with the actual.

**Review by E1337Kat — APPROVED** [review]
I think this is good. The tests all pass and I think I follow the logic well enough.

**Pull request merged**